consensus/src/consensus/inner.rs

```diff
@@ -0,0 +1,508 @@
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use crate::{
+    error::ConsensusError,
+    Consensus,
+    ConsensusParameters,
+    CreatePartialTransactionRequest,
+    DeserializedLedger,
+    DynLedger,
+    MemoryPool,
+    Tx,
+};
+use anyhow::*;
+use snarkos_storage::{
+    BlockStatus,
+    Digest,
+    ForkDescription,
+    SerialBlock,
+    SerialTransaction,
+    Storage,
+    VMRecord,
+    VMTransaction,
+};
+use snarkvm_dpc::{
+    testnet1::{
+        instantiated::{Components, InstantiatedDPC},
+        payload::Payload as RecordPayload,
+        transaction::amount::AleoAmount,
+        Record as DPCRecord,
+        TransactionKernel,
+    },
+    DPCScheme,
+};
+use snarkvm_posw::txids_to_roots;
+use snarkvm_utilities::{to_bytes, ToBytes};
+use std::io::Result as IoResult;
+use tokio::sync::mpsc;
+
+use snarkos_metrics::{self as metrics, misc::*};
+
+use rand::thread_rng;
+
+use super::message::{ConsensusMessage, CreateTransactionRequest, TransactionResponse};
+
+pub struct ConsensusInner {
+    pub public: Arc<Consensus>,
+    pub ledger: DynLedger,
+    pub memory_pool: MemoryPool,
+    pub storage: Arc<dyn Storage>,
+}
+
+impl ConsensusInner {
+    fn receive_transaction(&mut self, transaction: Box<SerialTransaction>) -> bool {
+        if transaction.value_balance.is_negative() {
+            error!("Received a transaction that was a coinbase transaction");
+            return false;
+        }
+        match self.verify_transaction(&*transaction) {
+            Ok(true) => (),
+            Ok(false) => {
+                error!("Received a transaction that was invalid");
+                return false;
+            }
+            Err(e) => {
+                error!(
+                    "failed to validate transaction -- note this does not mean the transaction was valid or invalid: {:?}",
+                    e
+                );
+                return false;
+            }
+        }
+        match self.memory_pool.insert(&self.ledger, *transaction) {
+            Ok(Some(digest)) => {
+                debug!("pushed transaction into memory pool: {:02X?}", &digest[..]);
+                true
+            }
+            Ok(None) => false,
+            Err(e) => {
+                error!("failed pushing transaction into memory pool: {:?}", e);
+                false
+            }
+        }
+    }
+
+    /// initialize genesis block if neccesary and catches up the chain
+    async fn init(&mut self) -> Result<()> {
+        let canon = self.storage.canon().await?;
+        if canon.block_height == 0 && canon.hash.len() == 0 {
```

Comment:
```suggestion
        if canon.block_height == 0 && canon.hash.is_empty() {
```
(before `clippy` has a chance to complain)